modules/catalog/src/main/java/org/apache/ignite/internal/catalog/CatalogServiceImpl.java

```diff
@@ -321,6 +341,89 @@ public CompletableFuture<Void> dropColumn(AlterTableDropColumnParams params) {
         });
     }
 
+    /** {@inheritDoc} */
+    @Override
+    public CompletableFuture<Void> alterColumn(AlterColumnParams params) {
+        return saveUpdate(catalog -> {
+            String schemaName = Objects.requireNonNullElse(params.schemaName(), CatalogService.PUBLIC);
+
+            SchemaDescriptor schema = Objects.requireNonNull(catalog.schema(schemaName), "No schema found: " + schemaName);
+
+            TableDescriptor table = schema.table(params.tableName());
+
+            if (table == null) {
+                throw new TableNotFoundException(schemaName, params.tableName());
+            }
+
+            String columnName = params.columnName();
+
+            Optional<TableColumnDescriptor> source = table.columns().stream().filter(desc -> desc.name().equals(columnName)).findFirst();
+
+            if (source.isEmpty()) {
+                throw new ColumnNotFoundException(columnName);
+            }
+
+            TableColumnDescriptor origin = source.get();
+
+            if (table.isPrimaryKeyColumn(origin.name())) {
+                if (params.notNull() != null) {
+                    throwUnsupportedDdl("Cannot change NOT NULL for the primary key column '{}'.", origin.name());
+                }
+
+                if (params.type() != null) {
+                    throwUnsupportedDdl("Cannot change data type for primary key column '{}'.", origin.name());
+                }
+            }
+
+            boolean nullable = !Objects.requireNonNullElse(params.notNull(), !origin.nullable());
+            DefaultValue defaultValue = Objects.requireNonNullElse(params.defaultValue(origin.type()), origin.defaultValue());
+            ColumnType type = Objects.requireNonNullElse(params.type(), origin.type());
+            int scale = Objects.requireNonNullElse(params.scale(), origin.scale());
+
+            boolean varLenType = type == ColumnType.STRING || type == ColumnType.BYTE_ARRAY;
+            int precision = varLenType ? origin.precision() : Objects.requireNonNullElse(params.precision(), origin.precision());
+            int length = varLenType ? Objects.requireNonNullElse(params.precision(), origin.length()) : origin.length();
```

Comment:
Why do we need a different logic for varlen types?
This logic is incorrect when you change fixlen to varlen, but this transition is forbidden now.
Why do you ignore precision and length from params? Maybe there should be an assert, that we don't expect any changes here?

Also, `Objects.requireNonNullElse(params.precision(), origin.length())` - I guess, here should be `params.length()`

Reply:
> Why do you ignore precision and length from params?

Currently I don't have length in params. After parsing we have RelDataType, which doesn't have a length.
But since there is a length in the TableColumnDescriptor - for VARCHAR/VARBINARY precision is used as the length, for the rest it is in precision.

If you think this is the wrong place for this transformation, I can add length to the params and set it when converting, WDYT?

Reply:
Added `length` to params.